REGRESSION (iOS 26): MediaRecorder produces corrupted video/audio files, resulting in some loss of content

jyavenard opened the pull request:
#### 812a4cd88295f97f134a61e9f7a3621530188f4c
<pre>
REGRESSION (iOS 26): MediaRecorder produces corrupted video/audio files, resulting in some loss of content
<a href="https://bugs.webkit.org/show_bug.cgi?id=299164">https://bugs.webkit.org/show_bug.cgi?id=299164</a>
<a href="https://rdar.apple.com/160982017">rdar://160982017</a>

Reviewed by Youenn Fablet.

In typical usage, a MediaRecorder is configured to generate segments of a finite size.
However, in the provided reproduction steps, no size are provided and the
audio recordings will be as long as the duration between the recording starting
and the recording being stopped.
When writing the last segment, if the AVAssetWriterInput reported not to be
ready for more data, we would abort in the expectation that another fetchData
would be called that would flush the remaining samples. However, this last fetchData
will never occur.

The MediaRecorderPrivateWriterAVFObjC made the incorrect assumptions that
the AVAssetWriterInput would always be accepting more content as we only feed
it compressed, interleaved sample. This assumption turned out to be false.

When closing the recording, we now pass to the writer all the pending samples
and write them asynchronously using the appropriate AVAssetWriterInput&apos;s APIs
that ensure writes are delayed until ready.
We now call AVAssetWriterInput as per its documentation suggests

Added test.

* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp:
(WebCore::MediaRecorderPrivateWriter::close):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
(WebCore::MediaRecorderPrivateWriterAVFObjC::MediaRecorderPrivateWriterAVFObjC):
(WebCore::MediaRecorderPrivateWriterAVFObjC::writeFrame):
(WebCore::MediaRecorderPrivateWriterAVFObjC::close):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp:
(WebCore::MediaRecorderPrivateWriterWebM::close):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h:

Canonical link: <a href="https://commits.webkit.org/300824@main">https://commits.webkit.org/300824@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3308e752e1b3966056d0ee9a6fe4a40d2d8c116b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123980 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43695 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34392 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130799 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76132 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d05723ac-2369-492b-b81c-c43ad43d7ad2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125857 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44419 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52289 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/336444fa-5604-406a-9bd2-5cf1bc81dc52) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126934 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35389 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110900 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74902 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4aee1ef2-28d9-44a5-b8ad-a5d3152b7b69) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34337 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29063 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74282 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105120 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29284 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/133473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50933 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38798 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/133473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51307 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107120 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/133473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26097 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47942 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47793 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50786 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56550 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50260 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53606 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51934 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
```diff
@@ -215,15 +220,74 @@ static inline void appendEndsPreviousSampleDurationMarker(AVAssetWriterInput *as
+        auto sample = samples.takeFirst();
+        if (sample->trackID() == m_audioTrackIndex) {
+            if (auto cmSample = toCMSampleBuffer(sample, m_audioDescription.get()))
+                audioSamples.append(*cmSample);
```
```suggestion
                audioSamples.append(WTFMove(*cmSample));
```

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
```diff
@@ -215,15 +220,74 @@ static inline void appendEndsPreviousSampleDurationMarker(AVAssetWriterInput *as
+            continue;
+        }
+        if (auto cmSample = toCMSampleBuffer(sample, m_videoDescription.get())) {
+            videoSamples.append(*cmSample);
```
```suggestion
            videoSamples.append(WTFMove(*cmSample));
```

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
```diff
@@ -215,15 +220,74 @@ static inline void appendEndsPreviousSampleDurationMarker(AVAssetWriterInput *as
+        GenericPromise::Producer producer;
+        promises.append(producer.promise());
+        [m_audioAssetWriterInput requestMediaDataWhenReadyOnQueue:queue.get() usingBlock:makeBlockPtr([producer = WTFMove(producer), samples = WTFMove(audioSamples), writerInput = m_audioAssetWriterInput]() mutable {
+            while ([writerInput isReadyForMoreMediaData]) {
```
What happens if `[writerInput isReadyForMoreMediaData]` returns false while samples is not emptied?
It seems we would never mark as finished writerInput and resolve producer.

I would tend to do something like:
```
while ([writerInput isReadyForMoreMediaData] && !samples.isEmpty())
    [writerInput appendSampleBuffer:samples.takeFirst().get()];
RELEASE_LOG_ERROR_IF(!samples.isEmpty()... );
[writerInput markAsFinished];
producer.resolve();
```

Ditto for video.

jyavenard replied:
We are in the callback; we get called again as soon as the value goes back to YES

jyavenard replied:
> [writerInput markAsFinished];
> producer.resolve();

that wouldn't work. We can only call markAsFinished once we've exhausted all the samples, and if `isReadyForMoreMediaData` returns `NO` then we need to wait for the next call from `requestMediaDataWhenReadyOnQueue` please see the documentation of AVAssetInputWriter https://developer.apple.com/documentation/avfoundation/avassetwriterinput/requestmediadatawhenready(on:using:)?language=objc

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
```diff
@@ -215,15 +220,74 @@ static inline void appendEndsPreviousSampleDurationMarker(AVAssetWriterInput *as
+        [m_videoAssetWriterInput requestMediaDataWhenReadyOnQueue:queue.get() usingBlock:makeBlockPtr([producer = WTFMove(producer), samples = WTFMove(audioSamples), writerInput = m_videoAssetWriterInput, hasAddedVideoFrame = m_hasAddedVideoFrame, endTime]() mutable {
+            while ([writerInput isReadyForMoreMediaData]) {
+                if (samples.size()) {
+                    [writerInput appendSampleBuffer:samples.takeFirst().get()];
```
Shouldn't we set hasAddedVideoFrame to true here?

jyavenard replied:
No. It set if we have a video frame above already

youennf commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
```diff
@@ -215,15 +220,74 @@ static inline void appendEndsPreviousSampleDurationMarker(AVAssetWriterInput *as
-        producer.resolve();
-    }).get()];
+    GenericPromise::all(promises)->whenSettled(m_waitingQueue, [producer = WTFMove(producer), writer = m_writer]() mutable {
+        [writer finishWritingWithCompletionHandler:makeBlockPtr([producer = WTFMove(producer)]() mutable {
```
```suggestion
        [writer finishWritingWithCompletionHandler:makeBlockPtr([producer = WTFMove(producer)] {
```
?

jyavenard replied:
It needs to be mutable Producer:resolve isn’t a const method

youennf commented on LayoutTests/http/wpt/mediarecorder/record-96KHz-long-mp4.html:
```diff
@@ -0,0 +1,61 @@
+        stream.getAudioTracks()[0].stop();
+
+        const mic96KHz = deviceFromLabel(devices, "Mock audio device 3");
+        stream = await navigator.mediaDevices.getUserMedia({ audio: { deviceId:mic96KHz.deviceId, sampleRate:96000 } })
```
Why do we need mic96KHz?

jyavenard replied:
I wanted to increase the number of samples so that the test can run for less time

jyavenard committed:
REGRESSION (iOS 26): MediaRecorder produces corrupted video/audio files, resulting in some loss of content
https://bugs.webkit.org/show_bug.cgi?id=299164
rdar://160982017

Reviewed by Youenn Fablet.

In typical usage, a MediaRecorder is configured to generate segments of a finite size.
However, in the provided reproduction steps, no size are provided and the
audio recordings will be as long as the duration between the recording starting
and the recording being stopped.
When writing the last segment, if the AVAssetWriterInput reported not to be
ready for more data, we would abort in the expectation that another fetchData
would be called that would flush the remaining samples. However, this last fetchData
will never occur.

The MediaRecorderPrivateWriterAVFObjC made the incorrect assumptions that
the AVAssetWriterInput would always be accepting more content as we only feed
it compressed, interleaved sample. This assumption turned out to be false.

When closing the recording, we now pass to the writer all the pending samples
and write them asynchronously using the appropriate AVAssetWriterInput's APIs
that ensure writes are delayed until ready.
We now call AVAssetWriterInput as per its documentation suggests

Added test.

* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.cpp:
(WebCore::MediaRecorderPrivateWriter::close):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateWriter.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.h:
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
(WebCore::MediaRecorderPrivateWriterAVFObjC::MediaRecorderPrivateWriterAVFObjC):
(WebCore::MediaRecorderPrivateWriterAVFObjC::writeFrame):
(WebCore::MediaRecorderPrivateWriterAVFObjC::close):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp:
(WebCore::MediaRecorderPrivateWriterWebM::close):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.h:

Canonical link: https://commits.webkit.org/300824@main
